File: r/sparklabs/batch/sales/dataframe/dq_data_quality.R
source("r/sparklabs/batch/sales/dataframe/utils.R")

duplicates <- where(
  count(groupBy(customers, customers$customer_id)),
  "count > 1"
)

if (count(duplicates) == 0) {
  cat("PASS: customer_id is unique\n")
} else {
  cat("FAIL: duplicate customer_id values found\n")
  showDF(duplicates, truncate = FALSE)
}

File: python/sparklabs/batch/sales/dq_data_quality.py
from pyspark.sql import functions as F

from sparklabs.batch.sales import utils


def run() -> None:
    duplicate_ids = (
        utils.customers.groupBy("customer_id")
        .count()
        .where(F.col("count") > 1)
    )
    duplicate_count = duplicate_ids.count()
    if duplicate_count == 0:
        print("PASS: customer_id is unique")
    else:
        print(f"FAIL: found {duplicate_count} duplicate customer_id values")
        duplicate_ids.show(truncate=False)


if __name__ == "__main__":
    run()
